### PR TITLE
fix: Stop assuming en0 is the primary ethernet interface

### LINKED ISSUE #1107: Sentry crashes if en0 is not the primary ethernet interface on macOS
### Environment

How do you use Sentry?
On-premise

Which SDK and version?
sentry-cocoa 7.0.3

### Steps to Reproduce

1. Change the primary ethernet interface to en1, or remove en0
2. Launch a macOS app using the Sentry SDK

Note: these steps involve deliberately changing the primary ethernet interface away from en0 if it is en0, but Apple recommends against assuming en0 exists at all and assuming it is the primary interface and we've seen this crash in our app with users on stock systems.

### Expected Result

Sentry doesn't crash.

### Actual Result

Sentry crashes with the following error:

```
ERROR: SentryCrashSysCtl.c (214): _Bool sentrycrashsysctl_getMacAddress(const char *const, char *const): Could not get interface index for en0: Device not configured
ERROR: SentryCrashMachineContext.c (230): void sentrycrashmc_resumeEnvironment(): thread_resume (00006707): (os/kern) invalid argument
[1] 4403 illegal hardware instruction
```

--- 

This issue occurs because `getDeviceAndAppHash` uses `sentrycrashsysctl_getMacAddress` to get the MAC address of the system, but assumes that the default interface is always on en0, when in fact, some systems do not have an en0:

From [Apple's sample code](https://developer.apple.com/library/archive/samplecode/GetPrimaryMACAddress/Listings/GetPrimaryMACAddress_GetPrimaryMACAddress_c.html#//apple_ref/doc/uid/DTS10000698-GetPrimaryMACAddress_GetPrimaryMACAddress_c-DontLinkElementID_3) on getting a MAC address:
```c
// Note that another option here would be:
// matchingDict = IOBSDMatching("en0");
// but en0: isn't necessarily the primary interface, especially on systems with multiple Ethernet ports.
```

I've already written a patch for this and will open a PR soon.

### PR DESCRIPTION
**Pull request by jamiebishop**
## :scroll: Description

Sentry currently crashes if en0 doesn't exist. This PR changes the way that MAC addresses are retrieved, to dynamically lookup the primary interface, using IOKit.

## :bulb: Motivation and Context

Resolves #1107 

## :pencil: Checklist

- [ ] I reviewed the submitted code
- [ ] I added tests to verify the changes
- [ ] I updated the docs if needed
- [ ] Review from the native team if needed
- [ ] No breaking changes

### CODE REVIEW

File: Sources/SentryCrash/Recording/Tools/SentryCrashIOKit.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
+//
+//  SentryCrashIOKit.h
+//  Sentry
+//
+//  Created by Jamie Bishop on 07/05/2021.
+//  Copyright © 2021 Sentry. All rights reserved.
+//
+
```
Comment (philipphofmann) [inline Sources/SentryCrash/Recording/Tools/SentryCrashIOKit.h:8]: `l`: We always remove these.
```suggestion
```

**Review by philipphofmann — COMMENTED** [review]
Hey, @jamiebishop thanks for opening up this PR. I just enabled CI for this PR and I see that most of the actions are failing. [IOKit](https://developer.apple.com/documentation/iokit) is only available on macOS 10.0+ and Mac Catalyst 13.0+. Either we keep both solutions and switch depending on the platform or maybe can you come up with a fix that works on all platforms?

For your info we use the LOGAF scale for marking comments in PRs:
* `l`: low
* `m`: medium
* `h`: high

**Comment by jamiebishop** [comment]
Hi @philipphofmann - thanks for taking a look at this. I'll get those copyright headers removed in a short while, but just on the failing tests - you can't get the MAC address on iOS since iOS 7 (and the old method just returned 0 for it), and it's the same on tvOS and watchOS.

In fact, the current code already makes use of UIDevice's identifierForVendor instead of the MAC address, if it's available: https://github.com/getsentry/sentry-cocoa/blob/cf7ae77743c6fdbd7d50cfd86389d2671dbc3c96/Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_System.m#L390.

So I think the best solution would be to update that code to add support for watchOS's [identifierForVendor](https://developer.apple.com/documentation/watchkit/wkinterfacedevice/3538473-identifierforvendor?language=objc) (watchOS 6.2+) and then moving the IOKit MAC address code to a macOS-only target.

**Comment by jamiebishop** [comment]
@philipphofmann do you have any thoughts on this approach?

**Comment by philipphofmann** [comment]
> @philipphofmann do you have any thoughts on this approach?

Sorry, @jamiebishop. I didn't get the time yet to take a look at your changes. I try to get back to you later. I did not forget.

**Comment by jamiebishop** [comment]
Hey @philipphofmann! I've fixed the tests and only #if'd out the new MAC address retrieval code on non-macOS platforms. Should be good for a review now.

**Comment by philipphofmann** [comment]
Thanks, @jamiebishop. Can you also please merge the latest master branch into your branch, please.

**Comment by philipphofmann** [comment]
And also please fix the failing CI.

**Comment by jamiebishop** [comment]
Thanks. Sorry, I can’t run CI as I haven’t contributed yet, which is why it’s falling. Will take a look at it soon

**Comment by jamiebishop** [comment]
I've merged in the latest master, and the failing CI should be fixed now. I don't like how heavy I had to be with the `#if`'s, but I don't really see a better way to do this.

**Review by BertsLifee — APPROVED** [review]

**Pull request closed without merging**